File: App/JsonParser.cs
using System.Text.Json;
using App.Models;

namespace App;

public static class JsonParser
{
    private static readonly JsonSerializerOptions Opts = new() { PropertyNameCaseInsensitive = true };
    
    public static IEnumerable<MockEndpoint> Parse(string json)
        => (
            JsonSerializer.Deserialize<Dictionary<string, JsonEndpoint>>(json,Opts)
            ?? new Dictionary<string, JsonEndpoint>()
        ).Select(endpoint => new MockEndpoint
        {
            Route = endpoint.Key,
            ResponseGenerator = endpoint.Value.ResponseGenerator,
            Response = endpoint.Value.Response,
            Methods = endpoint.Value.Method is not null ? [endpoint.Value.Method] : endpoint.Value.Methods ?? Enumerable.Empty<string>()
        });

    private readonly struct JsonEndpoint
    {
        public string? Method { get; init; }
        public string[]? Methods { get; init; }
        public string? Response { get; init; }
        public string? ResponseGenerator { get; init; }
    }
}

File: src/json_parser.rs
use serde::Deserialize;

#[derive(Deserialize, Debug)]
struct JsonEndpoint {
    method: Option<String>,
    methods: Option<Vec<String>>,
    response: Option<String>,
    response_generator: Option<String>,
}

#[derive(Debug, Clone)]
pub struct MockEndpoint {
    pub route: String,
    pub response_generator: Option<String>,
    pub response: Option<String>,
    pub methods: Vec<String>,
}

pub mod json_parser {
    use super::{JsonEndpoint, MockEndpoint};
    use std::collections::HashMap;

    pub fn parse(json: &str) -> Vec<MockEndpoint> {
        let endpoints: HashMap<String, JsonEndpoint> =
            serde_json::from_str(json).unwrap_or_default();

        endpoints
            .into_iter()
            .map(|(route, endpoint)| MockEndpoint {
                route,
                response_generator: endpoint.response_generator,
                response: endpoint.response,
                methods: match endpoint.method {
                    Some(method) => vec![method],
                    None => endpoint.methods.unwrap_or_default(),
                },
            })
            .collect()
    }
}
